### PR TITLE
feat(manipulation): add client-only Arm SDK for dimos shell

### PR DESCRIPTION
**Pull request by TomCC7**
## Contribution path

- Discussion: https://github.com/dimensionalOS/dimos/discussions/3923
- Scope: client-only motion conveniences over the existing manipulation RPCs.
- Generic Spec discovery and RPC completion landed in #4009.
- Independent perception startup fix: #4008. Typed pick/place changes have been dropped, not moved to a follow-up.
- closes DIM-1472

## Problem

Basic Python motion scripts must select a planning group, construct messages and mappings, and check planning and execution results by hand. The standalone manipulation runner duplicates runtime and client lifecycle already handled by `dimos run` and `dimos shell`.

## Solution

`Arm` is a client-side API for controlling one arm at a time. It binds a planning group once, so each call can use joint positions or XYZ values directly. In `dimos shell`, use the existing connection:

```python
from dimos.manipulation.sdk import Arm

arm = Arm.from_app(app)
```

| API | What it does | Returns |
|---|---|---|
| `Arm.from_app(app)` | Select the unique pose-capable arm, with or without a gripper. Use `group=` and `instance_name=` when selection is ambiguous. | `Arm` |
| `arm.info` | Inspect the group ID, joint order, frames, and gripper capability. | `PlanningGroupInfo` |
| `arm.state()` | Read the arm's state, including available joint presets. | `PlanningGroupState` |
| `arm.joints()` | Read a fresh joint-position array in `arm.info.joint_names` order. | NumPy array |
| `arm.pose()` | Read the current end-effector pose. | `PoseStamped` |
| `arm.move_joints(positions)` | Plan and execute a move to joint positions supplied as a list, tuple, or NumPy array. | `ExecutionResult` |
| `arm.move_pose(position, orientation=None)` | Plan and execute a move to world-frame XYZ. Optional orientation is an XYZW quaternion; omitting it preserves the current orientation. | `ExecutionResult` |
| `arm.move_linear(dx=0, dy=0, dz=0)` | Execute a straight world-frame translation through the existing motion primitive. The SDK adds no path-planning call. | `MoveResult` |
| `arm.set_gripper_position(position)` | Set normalized gripper travel: `0.0` closed, `1.0` open. | `CommandResult` |
| `arm.open_gripper()` / `arm.close_gripper()` | Open or close the selected arm's gripper. | `CommandResult` |
| `arm.home()` / `arm.move_to_preset(name)` | Move to a server-provided preset, using the same Home/Init values as Viser. | `ExecutionResult` |
| `arm.rpc` | Access explicit planning, preview, nonblocking execution, wait, and cancellation. | `ManipulationSpec` proxy |

Joint, pose, linear, and preset moves accept `speed_scale=` and `timeout=`. They wait for completed execution and raise `MotionError` on failure, retaining the original result in `error.result`. Positions and translations use metres; angular joints use radians.

Linear motion keeps the RPC's `check_collision=False` default. Pass `check_collision=True` to check the straight segment against the configured world. A timeout does not cancel motion.

Each `Arm` selects one group; this wrapper does not coordinate multi-arm motion. The module still has one shared pending plan and execution, so use one motion-commanding client per module. `Arm` borrows the connection: scripts use `Dimos.connect()`, and shell users reuse `app`.

This PR also completes the motion contract with typed preview/clear results and visualization discovery, removes the obsolete singular `plan_to_pose` RPC and unused helpers, and removes the standalone manipulation client runner. Manual control uses `dimos shell` with ordinary SDK imports and IPython completion/help.

Motion algorithms and pick/place behavior are unchanged. Generic discovery is in #4009; the perception startup fix is in #4008.

### Manual testing with the native MuJoCo window

Use a desktop terminal and the provisioned environment for this branch. Run only one blueprint on the default transport bus; stop an existing run first only when safe. If startup hits the known detector import issue, use the separate perception fix in #4008; it is deliberately not bundled here.

**Terminal 1: start the simulation with its native window.** No LLM credentials are needed.

```bash
source .venv/bin/activate
env -u MUJOCO_GL dimos --simulation run xarm-perception-sim --headless false
```

`--headless false` opens the native MuJoCo viewer. `env -u MUJOCO_GL` clears an earlier `egl` override. The global `--viewer` option controls Rerun, not the MuJoCo window. Viser controls are available at the URL printed during startup.

For a machine without a display, use this alternative:

```bash
MUJOCO_GL=egl dimos --simulation --viewer none run xarm-perception-sim --headless true
```

**Terminal 2: attach the shell** after modules finish starting, from the same worktree:

```bash
source .venv/bin/activate
dimos shell
```

Import and inspect without moving:

```python
from dimos.manipulation.sdk import Arm

arm = Arm.from_app(app)
arm.info
arm.joints()
arm.pose()
```

The shell already provides `app`; do not create a second connection. Use `arm.` followed by Tab and `arm.move_linear?` for help. These inspection commands do not move the arm.

**Try a small movement in simulation**, allowing room for a 1 cm vertical translation. Execute each command separately and inspect its result before continuing:

```python
arm.move_linear(dz=0.01, check_collision=True)
arm.move_linear(dz=-0.01, check_collision=True)
```

Watch the arm in MuJoCo. Each call should return a successful `MoveResult` with completed execution. If it raises, inspect the error and current state before issuing another command. A timeout does not mean the robot has stopped; the SDK does not retry or cancel automatically.

The [manual guide](https://github.com/dimensionalOS/dimos/blob/cc/feat/manip-client-sdk/docs/capabilities/manipulation/python_api.md) also covers joint/pose moves, restoring initial joints, gripper control, presets, and advanced RPCs.

**Finish:** enter `exit` in IPython. This disconnects the shell without stopping the runtime or cancelling motion. Stop the simulation with Ctrl-C in terminal 1 when finished. If no native window appears, inspect startup logs for GLFW/OpenGL errors; changing Rerun's `--viewer` setting will not enable it.

### Automated validation

```bash
python -m pytest dimos/manipulation/test_sdk.py dimos/cli/test_shell.py dimos/cli/test_cli_startup.py dimos/porcelain/test_dimos.py dimos/porcelain/test_remote_module_source.py dimos/manipulation/test_manipulation_unit.py dimos/manipulation/test_manipulation_spec.py dimos/manipulation/test_pick_and_place_unit.py dimos/manipulation/test_manipulation_monitor_preview.py dimos/manipulation/test_manipulation_primitives.py dimos/manipulation/test_plan_execution.py dimos/manipulation/test_execution_manager.py dimos/manipulation/visualization/test_operator.py dimos/codebase_checks/test_docs_branding.py -q
```

- 227 focused tests passed after the split, including 56 SDK tests, real IPython import/completion/help checks, shell lifecycle/startup, discovery, motion contracts/execution/preview, operator adapters, unchanged pick/place behavior, and documentation branding.
- Mypy passed on all four changed production Python files with `--follow-imports=silent`.
- Ruff, documentation links, and commit hooks passed.
- Earlier live headless simulation validation covered shell import/help, joint/pose/linear/restore moves, gripper commands, Home/Init reuse, and advanced preview/wait/cancel. A new client read live state after shell exit, confirming the runtime stayed up.
- No robot or simulation runtime was started or stopped during this PR split. Live tests above are historical, not rerun on the split head. No real hardware was tested; full CI remains to be checked on the new heads.

## AI assistance

Codex (GPT-5-based coding agent) implemented the SDK, tests, documentation, and review-driven PR split. The user directed the API and scope decisions. Draft for human review.

## Checklist

- [ ] I have read and approved the [CLA](https://github.com/dimensionalOS/dimos/blob/main/CLA.md).

### CODE REVIEW

File: docs/usage/cli.md [hunk 1]
```diff
@@ -225,6 +225,11 @@ The shell normally displays the CLI run ID and blueprint. Coordinators launched
 directly from Python have no run-registry metadata and are shown as
 `unregistered coordinator`.
 
+SDKs use ordinary Python imports and the existing `app`; the shell does not
+preload robot-specific interfaces. See the
+[manipulation walkthrough](/docs/capabilities/manipulation/python_api.md) for
+using `Arm`, inspecting methods, and testing motion interactively.
+
```
Comment (TomCC7) [inline docs/usage/cli.md]:
```suggestion
```

File: docs/usage/python-api.md [hunk 2]
```diff
@@ -49,10 +49,53 @@ app.GO2Connection.move(Twist(linear=(0, 0, 0), angular=(0, 0, -1)), duration=0.0
...
+accepts numeric sequences and NumPy arrays for joint and pose targets. Explicit
+planning and execution controls remain available through `arm.rpc`. See the
+[manipulation Python guide](/docs/capabilities/manipulation/python_api.md) for
+presets, failure handling, and simulation examples.
+
 ## Discovering modules and RPCs
 
 Discovery works in both local and remote mode:
 
+For a typed capability, use `app.get_module(MySpec)`. DimOS resolves the single
```
Comment (TomCC7) [inline docs/usage/python-api.md]: this can be a separate PR? should propose split and do a stacked pr to make boundary cleaner

File: dimos/manipulation/pick_and_place_spec.py
Comment (TomCC7) [inline dimos/manipulation/pick_and_place_spec.py:1]: how come we add a whole new spec file while not reducing the loc a lot in pick and place module itself? also do we mainly targetting pick and place modules any test/demos?

File: dimos/perception/experimental/object_scene_registration.py [hunk 3]
```diff
@@ -27,6 +27,10 @@
 from dimos.core.stream import In, Out
 from dimos.models.segmentation.edge_tam import BoxPromptImageSegmenter
 from dimos.models.segmentation.yoloe import YoloeBoxSegmenter
+
+# Resolve detector imports while the worker loads the module, before parallel
```
Comment (TomCC7) [inline dimos/perception/experimental/object_scene_registration.py]: this should be a separate pr as well

File: dimos/porcelain/dimos.py
Comment (TomCC7) [inline dimos/porcelain/dimos.py:1]: are changes here still 'required' if we already provide wrapped up sdk?

File: CONTEXT.md
Comment (TomCC7) [inline CONTEXT.md:1]: remove

File: docs/adr/0001-typed-manipulation-module-contracts.md
Comment (TomCC7) [inline docs/adr/0001-typed-manipulation-module-contracts.md:1]: remove

File: docs/usage/cli.md (hunk 1 above)
Comment (TomCC7) [reply docs/usage/cli.md]: Removed the SDK/manipulation paragraph from the generic CLI guide. The motion walkthrough stays in the dedicated manipulation Python guide.

File: docs/usage/python-api.md (hunk 2 above)
Comment (TomCC7) [reply docs/usage/python-api.md]: Split into draft #4009, containing only generic Spec-based discovery, its tests, and generic API documentation. #3944 now targets that branch, so the upper diff is motion SDK/contracts and shell documentation only. Discovery has 60 passing focused tests.

File: dimos/manipulation/pick_and_place_spec.py
Comment (TomCC7) [reply dimos/manipulation/pick_and_place_spec.py:1]: The extra LOC came from typed result contracts and agent formatting, not from extracting the pick/place algorithm. Following the agreed scope, I dropped the entire typed pick/place change from this PR, including its Spec, new tests, and walkthrough. PickAndPlaceModule and its existing unit tests now match main; there is no follow-up pick/place PR from this work. #3944 targets motion SDK use in dimos shell, with 56 SDK tests and manual motion instructions. No successful end-to-end pick/place cycle is claimed.

File: dimos/perception/experimental/object_scene_registration.py (hunk 3 above)
Comment (TomCC7) [reply dimos/perception/experimental/object_scene_registration.py]: Extracted into independent draft #4008, based on main. It contains only the detector import-order fix and its fresh-process regression. The regression now works without optional scikit-learn: verified green with sklearn unavailable, red after removing the eager imports, and green after restoring them. The startup/temporal suite passes 16 tests. These files are no longer changed by #3944.

File: dimos/porcelain/dimos.py
Comment (TomCC7) [reply dimos/porcelain/dimos.py:1]: Not inherently required for a wrapper: Arm could use a hardcoded module name. The current Arm.from_app() uses Spec discovery to avoid coupling to an implementation name, and other Python clients can reuse that capability. As agreed, this generic behavior is now isolated in #4009; #3944 builds on it without duplicating discovery in the SDK.

File: CONTEXT.md
Comment (TomCC7) [reply CONTEXT.md:1]: Removed CONTEXT.md from the PR.

File: docs/adr/0001-typed-manipulation-module-contracts.md
Comment (TomCC7) [reply docs/adr/0001-typed-manipulation-module-contracts.md:1]: Removed the ADR from the PR.

File: dimos/manipulation/manipulation_module.py [hunk 4]
```diff
@@ -1491,39 +1461,6 @@ def is_capable(group: PlanningGroup) -> bool:
             )
         return candidates[0]
 
-    def _lift_if_low(
```
Comment (TomCC7) [inline dimos/manipulation/manipulation_module.py:1485]: are these being used anywhere? or deadcode

File: dimos/manipulation/manipulation_module.py [hunk 5]
```diff
@@ -753,25 +747,6 @@ def solve_ik(
             self._record_error(f"IK failed: {result.status.name}{detail}")
         return result
 
-    @rpc
-    def plan_to_pose(self, pose: Pose, group_id: PlanningGroupID | None = None) -> bool:
```
Comment (TomCC7) [inline dimos/manipulation/manipulation_module.py:746]: why we removed this....

File: dimos/manipulation/planning/examples/manipulation_client.py
Comment (mustafab0) [inline dimos/manipulation/planning/examples/manipulation_client.py:1]: 😢 

Sad to see my friend go

Comment (TomCC7) [reply dimos/manipulation/planning/examples/manipulation_client.py:1]: or view it differently:
1. `dimos shell` is an ipython version of this repl, but serving the whole dimos rather than just manipulation
2. and sdk introduced here is just inspired by the convenience function you defined in the manip client :)

so nothing "go", they just transformed into a more generalized form (we can surely add nav sdk later)

File: dimos/manipulation/sdk.py
Comment (mustafab0) [inline dimos/manipulation/sdk.py:1]: Let's move this to 

`dimos/manipulation/sdk`

to keep our manipulation boundary clear for now

**Review by mustafab0 — CHANGES_REQUESTED** [review]

File: dimos/manipulation/sdk.py
Comment (mustafab0) [reply dimos/manipulation/sdk.py]: oh I think there is a confusion. 

I do think sdk folder is a good idea. It just lives 

`dimos/manipulation/sdk/sdk.py`

like so, for now.

**Review by mustafab0 — CHANGES_REQUESTED** [review]

**Review by mustafab0 — APPROVED** [review]
